src/pattern.rs

```diff
@@ -144,10 +144,22 @@ where
     M: Metadata<L>,
 {
     fn search(&self, egraph: &EGraph<L, M>) -> Vec<SearchMatches> {
-        egraph
-            .classes()
-            .filter_map(|e| self.search_eclass(egraph, e.id))
-            .collect()
+	match &self.ast {
+	    PatternAst::ENode(e) => {
+		if let Some(ids) = egraph.sigtoclasses.get(&(e.op.clone(), e.children.len())) {
+		    ids.iter().filter_map(|id| self.search_eclass(egraph, *id))
+			.collect()
+		} else {
+		    vec![]
+		}
+	    },
+	    PatternAst::Var(v) => {
+		egraph
+		    .classes()
+		    .filter_map(|e| self.search_eclass(egraph, e.id))
+		    .collect()
+	    }
+	}
```

Comment:
Could be put this logic in `machine` instead? It should make it even faster, because it will apply on recursive searches as well.

Reply:
ah nvm, I think I see. It makes more sense here, since machine will only look at one eclass at a time.

Reply:
Good idea, I'll try that

Reply:
I might be able to do that if I make the classes_by_op refer to a set

Reply:
I worry that might be slower (at least for some eclasses), because the machine could either just look at the eclass instead

Reply:
I think that's probably the case based on my other experiments